### DISCUSSION TITLE
Problem while building srsRAN from source

### DISCUSSION THREAD

**Discussion by Monsieur-ugur** [srsran/srsRAN_Project#684] [General Help] [answered]
Good morning, I encounter errors while building the make in the build directory.
I am not working on a VM, my processor is the " Intel Atom(R) x6425RE Processor @ 1.90GHz", it supports sse4.1, sse1, and sse2 and I am on ubuntu 20.04.6, here is a list a the supported flag :

<img width="949" alt="tmp_github" src="https://github.com/srsran/srsRAN_Project/assets/116889189/a8bbb90a-5362-4a32-aec3-aeea97f52d11">

someone had a similar error but he was on a VM, here is a link : https://github.com/srsran/srsRAN_Project/discussions/151

The result from the cmake output : 

sudo cmake
../
-- The C compiler identification is GNU 10.5.0
-- The CXX compiler identification is GNU 9.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build type not specified: defaulting to Release.
-- Performing Test HAS_MAYBE_UNINITIALIZED
-- Performing Test HAS_MAYBE_UNINITIALIZED - Success
-- Performing Test HAS_STRINGOP_OVERFLOW
-- Performing Test HAS_STRINGOP_OVERFLOW - Success
-- Performing Test HAVE_NON_VIRTUAL_DTOR
-- Performing Test HAVE_NON_VIRTUAL_DTOR - Success
-- Performing Test HAVE_SUGGEST_OVERRIDE
-- Performing Test HAVE_SUGGEST_OVERRIDE - Success
-- Performing Test HAVE_SHADOW
-- Performing Test HAVE_SHADOW - Success
-- Assertion level set to NORMAL
-- Could NOT find libdw (missing: LIBDW_LIBRARY LIBDW_INCLUDE_DIR)
-- Found libbfd: /usr/lib/x86_64-linux-gnu/libbfd.so
-- Could NOT find libdwarf (missing: LIBDWARF_LIBRARY LIBDWARF_INCLUDE_DIR LIBELF_LIBRARY LIBELF_INCLUDE_DIR)
-- Found Backward: /home/ubuntu/5G_testbed/srsRAN_Project/cmake/modules
-- Building with backward-cpp support
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
-- Checking for module 'mbedtls'
--   No package 'mbedtls' found
-- MBEDTLS LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so
-- MBEDTLS STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.a
-- MBEDTLS INCLUDE DIRS: /usr/include
-- Found MbedTLS: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so
-- Checking for module 'fftw3f >= 3.0'
--   Found fftw3f , version 3.3.8
-- FFTW3F LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- FFTW3F STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.a
-- FFTW3F INCLUDE DIRS: /usr/include
-- Found FFTW3F: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- Found GTest: /usr/lib/x86_64-linux-gnu/libgtest.a
-- Found Threads: TRUE
-- UHD LIBRARIES /usr/local/lib/libuhd.so
-- UHD INCLUDE DIRS /usr/local/include
-- Found UHD: /usr/local/lib/libuhd.so
-- Checking for module 'yaml-cpp'
--   Found yaml-cpp, version 0.6.2
-- YAMLCPP LIBRARIES: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so
-- YAMLCPP INCLUDE DIRS: /usr/include
-- Found YAMLCPP: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so
-- Performing Test HAVE_SSE
-- Performing Test HAVE_SSE - Success
-- SSE4.1 is enabled - target CPU must support it
-- Performing Test HAVE_AVX
-- Performing Test HAVE_AVX - Failed
-- Performing Test HAVE_AVX2
-- Performing Test HAVE_AVX2 - Failed
-- Performing Test HAVE_FMA
-- Performing Test HAVE_FMA - Failed
-- Performing Test HAVE_AVX512
-- Performing Test HAVE_AVX512 - Failed
-- Performing Test HAVE_MARCH
-- Performing Test HAVE_MARCH - Success
-- Enabling ccache for C
-- Enabling ccache for CXX
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.17") found components: doxygen missing components: dot
-- Checking for module 'sctp'
--   No package 'sctp' found
-- SCTP LIBRARIES: /usr/lib/x86_64-linux-gnu/libsctp.so
-- SCTP INCLUDE DIRS: /usr/include
-- Found SCTP: /usr/lib/x86_64-linux-gnu/libsctp.so
-- FFT_LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- Building srsRAN version 24.04.0
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ubuntu/5G_testbed/srsRAN_Project/build

I have all of the dependencies installed (I have double-checked)

My errors while building the "make -j $(nproc)" are the following  :

[ 30%] Building CXX object lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/channel_equalizer_generic_impl.cpp.o
In file included from /usr/lib/gcc/x86_64-linux-gnu/9/include/immintrin.h:107,
                 from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/../../../srsvec/simd.h:36,
                 from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/equalize_mmse_1xn.h:28,
                 from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/channel_equalizer_generic_impl.cpp:27:
/usr/lib/gcc/x86_64-linux-gnu/9/include/fmaintrin.h: In function ‘srsran::simd_f_t srsran::srsran_simd_f_fma(srsran::simd_f_t, srsran::simd_f_t, srsran::simd_f_t)’:
/usr/lib/gcc/x86_64-linux-gnu/9/include/fmaintrin.h:55:1: error: inlining failed in call to always_inline ‘__m128 _mm_fmadd_ps(__m128, __m128, __m128)’: target specific option mismatch
   55 | _mm_fmadd_ps (__m128 __A, __m128 __B, __m128 __C)
      | ^~~~~~~~~~~~
In file included from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/equalize_mmse_1xn.h:28,
                 from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/channel_equalizer_generic_impl.cpp:27:
/home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/../../../srsvec/simd.h:395:22: note: called from here
  395 |   return _mm_fmadd_ps(a, b, acc);
      |          ~~~~~~~~~~~~^~~~~~~~~~~
make[2]: *** [lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/build.make:63: lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/channel_equalizer_generic_impl.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:11743: lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/all] Error 2
make: *** [Makefile:141: all] Error 2

Wishing you a nice day.

**Comment by Omar-D** [comment]
I am working on Gentoo Linux last version, i am getting the same issue.

i don't think that the problem is in cmake because i have tried multiple versions of cmake to compile it but everytime facing the same problem so i think the problem exactly is in the source code or maybe there is something we miss.

i will try to solve it, if i get the solution i will give it to you!
i hope you are gonna do the same

Wishing you a nice day.
Thanks in advance

**Comment by Omar-D** [comment, marked as answer]
> Good morning, I encounter errors while building the make in the build directory. I am not working on a VM, my processor is the " Intel Atom(R) x6425RE Processor @ 1.90GHz", it supports sse4.1, sse1, and sse2 and I am on ubuntu 20.04.6, here is a list a the supported flag :
> <img alt="tmp_github" width="949" src="https://private-user-images.githubusercontent.com/116889189/341345099-a8bbb90a-5362-4a32-aec3-aeea97f52d11.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5MjIxNjUsIm5iZiI6MTcxODkyMTg2NSwicGF0aCI6Ii8xMTY4ODkxODkvMzQxMzQ1MDk5LWE4YmJiOTBhLTUzNjItNGEzMi1hZWMzLWFlZWE5N2Y1MmQxMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMFQyMjE3NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZGNmNzYwNzhjN2FiYjRjNWM4MjllNTgzNDIxZWJlMjdhNzU2YTg1OTQxZGM0ZTI4NzRmZTFhOGU4NzdmNTAwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.nq8TfHekfg-03dmRfQIcnwaGbjW2_Kk5mXOITcxLtBo">
> 
> someone had a similar error but he was on a VM, here is a link : #151
> 
> The result from the cmake output :
> 
> sudo cmake ../ -- The C compiler identification is GNU 10.5.0 -- The CXX compiler identification is GNU 9.4.0 -- Check for working C compiler: /usr/bin/cc -- Check for working C compiler: /usr/bin/cc -- works -- Detecting C compiler ABI info -- Detecting C compiler ABI info - done -- Detecting C compile features -- Detecting C compile features - done -- Check for working CXX compiler: /usr/bin/c++ -- Check for working CXX compiler: /usr/bin/c++ -- works -- Detecting CXX compiler ABI info -- Detecting CXX compiler ABI info - done -- Detecting CXX compile features -- Detecting CXX compile features - done -- Build type not specified: defaulting to Release. -- Performing Test HAS_MAYBE_UNINITIALIZED -- Performing Test HAS_MAYBE_UNINITIALIZED - Success -- Performing Test HAS_STRINGOP_OVERFLOW -- Performing Test HAS_STRINGOP_OVERFLOW - Success -- Performing Test HAVE_NON_VIRTUAL_DTOR -- Performing Test HAVE_NON_VIRTUAL_DTOR - Success -- Performing Test HAVE_SUGGEST_OVERRIDE -- Performing Test HAVE_SUGGEST_OVERRIDE - Success -- Performing Test HAVE_SHADOW -- Performing Test HAVE_SHADOW - Success -- Assertion level set to NORMAL -- Could NOT find libdw (missing: LIBDW_LIBRARY LIBDW_INCLUDE_DIR) -- Found libbfd: /usr/lib/x86_64-linux-gnu/libbfd.so -- Could NOT find libdwarf (missing: LIBDWARF_LIBRARY LIBDWARF_INCLUDE_DIR LIBELF_LIBRARY LIBELF_INCLUDE_DIR) -- Found Backward: /home/ubuntu/5G_testbed/srsRAN_Project/cmake/modules -- Building with backward-cpp support -- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") -- Checking for module 'mbedtls' -- No package 'mbedtls' found -- MBEDTLS LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so -- MBEDTLS STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.a -- MBEDTLS INCLUDE DIRS: /usr/include -- Found MbedTLS: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so -- Checking for module 'fftw3f >= 3.0' -- Found fftw3f , version 3.3.8 -- FFTW3F LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so -- FFTW3F STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.a -- FFTW3F INCLUDE DIRS: /usr/include -- Found FFTW3F: /usr/lib/x86_64-linux-gnu/libfftw3f.so -- Found GTest: /usr/lib/x86_64-linux-gnu/libgtest.a -- Found Threads: TRUE -- UHD LIBRARIES /usr/local/lib/libuhd.so -- UHD INCLUDE DIRS /usr/local/include -- Found UHD: /usr/local/lib/libuhd.so -- Checking for module 'yaml-cpp' -- Found yaml-cpp, version 0.6.2 -- YAMLCPP LIBRARIES: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so -- YAMLCPP INCLUDE DIRS: /usr/include -- Found YAMLCPP: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so -- Performing Test HAVE_SSE -- Performing Test HAVE_SSE - Success -- SSE4.1 is enabled - target CPU must support it -- Performing Test HAVE_AVX -- Performing Test HAVE_AVX - Failed -- Performing Test HAVE_AVX2 -- Performing Test HAVE_AVX2 - Failed -- Performing Test HAVE_FMA -- Performing Test HAVE_FMA - Failed -- Performing Test HAVE_AVX512 -- Performing Test HAVE_AVX512 - Failed -- Performing Test HAVE_MARCH -- Performing Test HAVE_MARCH - Success -- Enabling ccache for C -- Enabling ccache for CXX -- Found Doxygen: /usr/bin/doxygen (found version "1.8.17") found components: doxygen missing components: dot -- Checking for module 'sctp' -- No package 'sctp' found -- SCTP LIBRARIES: /usr/lib/x86_64-linux-gnu/libsctp.so -- SCTP INCLUDE DIRS: /usr/include -- Found SCTP: /usr/lib/x86_64-linux-gnu/libsctp.so -- FFT_LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so -- Building srsRAN version 24.04.0 -- Configuring done -- Generating done -- Build files have been written to: /home/ubuntu/5G_testbed/srsRAN_Project/build
> 
> I have all of the dependencies installed (I have double-checked)
> 
> My errors while building the "make -j $(nproc)" are the following :
> 
> [ 30%] Building CXX object lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/channel_equalizer_generic_impl.cpp.o In file included from /usr/lib/gcc/x86_64-linux-gnu/9/include/immintrin.h:107, from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/../../../srsvec/simd.h:36, from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/equalize_mmse_1xn.h:28, from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/channel_equalizer_generic_impl.cpp:27: /usr/lib/gcc/x86_64-linux-gnu/9/include/fmaintrin.h: In function ‘srsran::simd_f_t srsran::srsran_simd_f_fma(srsran::simd_f_t, srsran::simd_f_t, srsran::simd_f_t)’: /usr/lib/gcc/x86_64-linux-gnu/9/include/fmaintrin.h:55:1: error: inlining failed in call to always_inline ‘__m128 _mm_fmadd_ps(__m128, __m128, __m128)’: target specific option mismatch 55 | _mm_fmadd_ps (__m128 __A, __m128 __B, __m128 __C) | ^~~~~~~~~~~~ In file included from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/equalize_mmse_1xn.h:28, from /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/channel_equalizer_generic_impl.cpp:27: /home/ubuntu/5G_testbed/srsRAN_Project/lib/phy/upper/equalization/../../../srsvec/simd.h:395:22: note: called from here 395 | return _mm_fmadd_ps(a, b, acc); | ~~~~~~~~~~~~^~~~~~~~~~~ make[2]: *** [lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/build.make:63: lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/channel_equalizer_generic_impl.cpp.o] Error 1 make[1]: *** [CMakeFiles/Makefile2:11743: lib/phy/upper/equalization/CMakeFiles/srsran_channel_equalizer.dir/all] Error 2 make: *** [Makefile:141: all] Error 2
> 
> Wishing you a nice day.

i found the solution, here is it:

`cmake ../ -DENABLE_EXPORT=ON -DENABLE_ZEROMQ=ON -DAUTO_DETECT_ISA=OFF`

`make -jCORE_NUMBER`

`make install -jCORE_NUMBER`

it was enough to solve the problem

Thanks in advance!

**Reply by dvdgrgrtt** [reply]
Hi!
You can also try with `-DAUTO_DETECT_ISA=ON -DENABLE_SSE=OFF -DENABLE_AVX=OFF -DENABLE_AVX2=OFF -DENABLE_FMA=OFF -DENABLE_AVX512=OFF` and then try to activate the different extensions one by one (SSE must be ON to activate any of the others), to see which one fails.
In this way, you get at least some SIMD support.

**Reply by Omar-D** [reply]
Hi!

Thanks for inform us, i appreciate that.
Wish you have a nice day

**Comment by Monsieur-ugur** [comment]
Thank you @Omar-D, I wish you a nice day. 
I am closing the discussion.

**Reply by Omar-D** [reply]
You are welcome brother!
